// @suilend/springsui-sdk/_generated/liquid_staking/weight/functions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.new_ = new_;
exports.collectFees = collectFees;
exports.init = init;
exports.updateFees = updateFees;
exports.addToRegistry = addToRegistry;
exports.adminCap = adminCap;
exports.eject = eject;
exports.rebalance = rebalance;
exports.setValidatorAddressesAndWeights = setValidatorAddressesAndWeights;
exports.weightHookId = weightHookId;
const __1 = require("..");
const util_1 = require("../../_framework/util");
function new_(tx, typeArg, adminCap) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::weight::new`,
        typeArguments: [typeArg],
        arguments: [(0, util_1.obj)(tx, adminCap)],
    });
}
function collectFees(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::weight::collect_fees`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.self),
            (0, util_1.obj)(tx, args.weightHookAdminCap),
            (0, util_1.obj)(tx, args.liquidStakingInfo),
            (0, util_1.obj)(tx, args.systemState),
        ],
    });
}
function init(tx, otw) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::weight::init`,
        arguments: [(0, util_1.obj)(tx, otw)],
    });
}
function updateFees(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::weight::update_fees`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.self),
            (0, util_1.obj)(tx, args.weightHookAdminCap),
            (0, util_1.obj)(tx, args.liquidStakingInfo),
            (0, util_1.obj)(tx, args.feeConfig),
        ],
    });
}
function addToRegistry(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::weight::add_to_registry`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.self),
            (0, util_1.obj)(tx, args.registry),
            (0, util_1.obj)(tx, args.liquidStakingInfo),
        ],
    });
}
function adminCap(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::weight::admin_cap`,
        typeArguments: [typeArg],
        arguments: [(0, util_1.obj)(tx, args.self), (0, util_1.obj)(tx, args.weightHookAdminCap)],
    });
}
function eject(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::weight::eject`,
        typeArguments: [typeArg],
        arguments: [(0, util_1.obj)(tx, args.self), (0, util_1.obj)(tx, args.adminCap)],
    });
}
function rebalance(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::weight::rebalance`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.self),
            (0, util_1.obj)(tx, args.systemState),
            (0, util_1.obj)(tx, args.liquidStakingInfo),
        ],
    });
}
function setValidatorAddressesAndWeights(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::weight::set_validator_addresses_and_weights`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.self),
            (0, util_1.obj)(tx, args.weightHookAdminCap),
            (0, util_1.obj)(tx, args.validatorAddressesAndWeights),
        ],
    });
}
function weightHookId(tx, self) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::weight::weight_hook_id`,
        arguments: [(0, util_1.obj)(tx, self)],
    });
}
