// @suilend/springsui-sdk/_generated/liquid_staking/weight/structs.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeightHookAdminCap = exports.WeightHook = exports.WEIGHT = exports.RegistryInfo = exports.CreateEvent = void 0;
exports.isCreateEvent = isCreateEvent;
exports.isRegistryInfo = isRegistryInfo;
exports.isWEIGHT = isWEIGHT;
exports.isWeightHook = isWeightHook;
exports.isWeightHookAdminCap = isWeightHookAdminCap;
const structs_1 = require("../../_dependencies/source/0x1/type-name/structs");
const structs_2 = require("../../_dependencies/source/0x2/bag/structs");
const structs_3 = require("../../_dependencies/source/0x2/object/structs");
const structs_4 = require("../../_dependencies/source/0x2/vec-map/structs");
const reified_1 = require("../../_framework/reified");
const util_1 = require("../../_framework/util");
const index_1 = require("../index");
const structs_5 = require("../liquid-staking/structs");
const structs_6 = require("../version/structs");
const bcs_1 = require("@mysten/sui/bcs");
const utils_1 = require("@mysten/sui/utils");
/* ============================== CreateEvent =============================== */
function isCreateEvent(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V5}::weight::CreateEvent`;
}
class CreateEvent {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = CreateEvent.$typeName;
        this.$isPhantom = CreateEvent.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(CreateEvent.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.typename = fields.typename;
        this.weightHookId = fields.weightHookId;
        this.weightHookAdminCapId = fields.weightHookAdminCapId;
    }
    static reified() {
        return {
            typeName: CreateEvent.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(CreateEvent.$typeName, ...[]),
            typeArgs: [],
            isPhantom: CreateEvent.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => CreateEvent.fromFields(fields),
            fromFieldsWithTypes: (item) => CreateEvent.fromFieldsWithTypes(item),
            fromBcs: (data) => CreateEvent.fromBcs(data),
            bcs: CreateEvent.bcs,
            fromJSONField: (field) => CreateEvent.fromJSONField(field),
            fromJSON: (json) => CreateEvent.fromJSON(json),
            fromSuiParsedData: (content) => CreateEvent.fromSuiParsedData(content),
            fromSuiObjectData: (content) => CreateEvent.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return CreateEvent.fetch(client, id); }),
            new: (fields) => {
                return new CreateEvent([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return CreateEvent.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(CreateEvent.reified());
    }
    static get p() {
        return CreateEvent.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("CreateEvent", {
            typename: structs_1.TypeName.bcs,
            weight_hook_id: structs_3.ID.bcs,
            weight_hook_admin_cap_id: structs_3.ID.bcs,
        });
    }
    static fromFields(fields) {
        return CreateEvent.reified().new({
            typename: (0, reified_1.decodeFromFields)(structs_1.TypeName.reified(), fields.typename),
            weightHookId: (0, reified_1.decodeFromFields)(structs_3.ID.reified(), fields.weight_hook_id),
            weightHookAdminCapId: (0, reified_1.decodeFromFields)(structs_3.ID.reified(), fields.weight_hook_admin_cap_id),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isCreateEvent(item.type)) {
            throw new Error("not a CreateEvent type");
        }
        return CreateEvent.reified().new({
            typename: (0, reified_1.decodeFromFieldsWithTypes)(structs_1.TypeName.reified(), item.fields.typename),
            weightHookId: (0, reified_1.decodeFromFieldsWithTypes)(structs_3.ID.reified(), item.fields.weight_hook_id),
            weightHookAdminCapId: (0, reified_1.decodeFromFieldsWithTypes)(structs_3.ID.reified(), item.fields.weight_hook_admin_cap_id),
        });
    }
    static fromBcs(data) {
        return CreateEvent.fromFields(CreateEvent.bcs.parse(data));
    }
    toJSONField() {
        return {
            typename: this.typename.toJSONField(),
            weightHookId: this.weightHookId,
            weightHookAdminCapId: this.weightHookAdminCapId,
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return CreateEvent.reified().new({
            typename: (0, reified_1.decodeFromJSONField)(structs_1.TypeName.reified(), field.typename),
            weightHookId: (0, reified_1.decodeFromJSONField)(structs_3.ID.reified(), field.weightHookId),
            weightHookAdminCapId: (0, reified_1.decodeFromJSONField)(structs_3.ID.reified(), field.weightHookAdminCapId),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== CreateEvent.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return CreateEvent.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isCreateEvent(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a CreateEvent object`);
        }
        return CreateEvent.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" || !isCreateEvent(data.bcs.type)) {
                throw new Error(`object at is not a CreateEvent object`);
            }
            return CreateEvent.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return CreateEvent.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching CreateEvent object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isCreateEvent(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a CreateEvent object`);
            }
            return CreateEvent.fromSuiObjectData(res.data);
        });
    }
}
exports.CreateEvent = CreateEvent;
CreateEvent.$typeName = `${index_1.PKG_V5}::weight::CreateEvent`;
CreateEvent.$numTypeParams = 0;
CreateEvent.$isPhantom = [];
/* ============================== RegistryInfo =============================== */
function isRegistryInfo(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V5}::weight::RegistryInfo`;
}
class RegistryInfo {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = RegistryInfo.$typeName;
        this.$isPhantom = RegistryInfo.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(RegistryInfo.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.weightHookId = fields.weightHookId;
    }
    static reified() {
        return {
            typeName: RegistryInfo.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(RegistryInfo.$typeName, ...[]),
            typeArgs: [],
            isPhantom: RegistryInfo.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => RegistryInfo.fromFields(fields),
            fromFieldsWithTypes: (item) => RegistryInfo.fromFieldsWithTypes(item),
            fromBcs: (data) => RegistryInfo.fromBcs(data),
            bcs: RegistryInfo.bcs,
            fromJSONField: (field) => RegistryInfo.fromJSONField(field),
            fromJSON: (json) => RegistryInfo.fromJSON(json),
            fromSuiParsedData: (content) => RegistryInfo.fromSuiParsedData(content),
            fromSuiObjectData: (content) => RegistryInfo.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return RegistryInfo.fetch(client, id); }),
            new: (fields) => {
                return new RegistryInfo([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return RegistryInfo.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(RegistryInfo.reified());
    }
    static get p() {
        return RegistryInfo.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("RegistryInfo", {
            weight_hook_id: structs_3.ID.bcs,
        });
    }
    static fromFields(fields) {
        return RegistryInfo.reified().new({
            weightHookId: (0, reified_1.decodeFromFields)(structs_3.ID.reified(), fields.weight_hook_id),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isRegistryInfo(item.type)) {
            throw new Error("not a RegistryInfo type");
        }
        return RegistryInfo.reified().new({
            weightHookId: (0, reified_1.decodeFromFieldsWithTypes)(structs_3.ID.reified(), item.fields.weight_hook_id),
        });
    }
    static fromBcs(data) {
        return RegistryInfo.fromFields(RegistryInfo.bcs.parse(data));
    }
    toJSONField() {
        return {
            weightHookId: this.weightHookId,
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return RegistryInfo.reified().new({
            weightHookId: (0, reified_1.decodeFromJSONField)(structs_3.ID.reified(), field.weightHookId),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== RegistryInfo.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return RegistryInfo.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isRegistryInfo(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a RegistryInfo object`);
        }
        return RegistryInfo.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" ||
                !isRegistryInfo(data.bcs.type)) {
                throw new Error(`object at is not a RegistryInfo object`);
            }
            return RegistryInfo.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return RegistryInfo.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching RegistryInfo object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isRegistryInfo(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a RegistryInfo object`);
            }
            return RegistryInfo.fromSuiObjectData(res.data);
        });
    }
}
exports.RegistryInfo = RegistryInfo;
RegistryInfo.$typeName = `${index_1.PKG_V5}::weight::RegistryInfo`;
RegistryInfo.$numTypeParams = 0;
RegistryInfo.$isPhantom = [];
/* ============================== WEIGHT =============================== */
function isWEIGHT(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V1}::weight::WEIGHT`;
}
class WEIGHT {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = WEIGHT.$typeName;
        this.$isPhantom = WEIGHT.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(WEIGHT.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.dummyField = fields.dummyField;
    }
    static reified() {
        return {
            typeName: WEIGHT.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(WEIGHT.$typeName, ...[]),
            typeArgs: [],
            isPhantom: WEIGHT.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => WEIGHT.fromFields(fields),
            fromFieldsWithTypes: (item) => WEIGHT.fromFieldsWithTypes(item),
            fromBcs: (data) => WEIGHT.fromBcs(data),
            bcs: WEIGHT.bcs,
            fromJSONField: (field) => WEIGHT.fromJSONField(field),
            fromJSON: (json) => WEIGHT.fromJSON(json),
            fromSuiParsedData: (content) => WEIGHT.fromSuiParsedData(content),
            fromSuiObjectData: (content) => WEIGHT.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return WEIGHT.fetch(client, id); }),
            new: (fields) => {
                return new WEIGHT([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return WEIGHT.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(WEIGHT.reified());
    }
    static get p() {
        return WEIGHT.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("WEIGHT", {
            dummy_field: bcs_1.bcs.bool(),
        });
    }
    static fromFields(fields) {
        return WEIGHT.reified().new({
            dummyField: (0, reified_1.decodeFromFields)("bool", fields.dummy_field),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isWEIGHT(item.type)) {
            throw new Error("not a WEIGHT type");
        }
        return WEIGHT.reified().new({
            dummyField: (0, reified_1.decodeFromFieldsWithTypes)("bool", item.fields.dummy_field),
        });
    }
    static fromBcs(data) {
        return WEIGHT.fromFields(WEIGHT.bcs.parse(data));
    }
    toJSONField() {
        return {
            dummyField: this.dummyField,
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return WEIGHT.reified().new({
            dummyField: (0, reified_1.decodeFromJSONField)("bool", field.dummyField),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== WEIGHT.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return WEIGHT.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isWEIGHT(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a WEIGHT object`);
        }
        return WEIGHT.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" || !isWEIGHT(data.bcs.type)) {
                throw new Error(`object at is not a WEIGHT object`);
            }
            return WEIGHT.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return WEIGHT.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching WEIGHT object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isWEIGHT(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a WEIGHT object`);
            }
            return WEIGHT.fromSuiObjectData(res.data);
        });
    }
}
exports.WEIGHT = WEIGHT;
WEIGHT.$typeName = `${index_1.PKG_V1}::weight::WEIGHT`;
WEIGHT.$numTypeParams = 0;
WEIGHT.$isPhantom = [];
/* ============================== WeightHook =============================== */
function isWeightHook(type) {
    type = (0, util_1.compressSuiType)(type);
    return type.startsWith(`${index_1.PKG_V1}::weight::WeightHook` + "<");
}
class WeightHook {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = WeightHook.$typeName;
        this.$isPhantom = WeightHook.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(WeightHook.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.id = fields.id;
        this.validatorAddressesAndWeights = fields.validatorAddressesAndWeights;
        this.totalWeight = fields.totalWeight;
        this.adminCap = fields.adminCap;
        this.version = fields.version;
        this.extraFields = fields.extraFields;
    }
    static reified(P) {
        return {
            typeName: WeightHook.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(WeightHook.$typeName, ...[(0, reified_1.extractType)(P)]),
            typeArgs: [(0, reified_1.extractType)(P)],
            isPhantom: WeightHook.$isPhantom,
            reifiedTypeArgs: [P],
            fromFields: (fields) => WeightHook.fromFields(P, fields),
            fromFieldsWithTypes: (item) => WeightHook.fromFieldsWithTypes(P, item),
            fromBcs: (data) => WeightHook.fromBcs(P, data),
            bcs: WeightHook.bcs,
            fromJSONField: (field) => WeightHook.fromJSONField(P, field),
            fromJSON: (json) => WeightHook.fromJSON(P, json),
            fromSuiParsedData: (content) => WeightHook.fromSuiParsedData(P, content),
            fromSuiObjectData: (content) => WeightHook.fromSuiObjectData(P, content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return WeightHook.fetch(client, P, id); }),
            new: (fields) => {
                return new WeightHook([(0, reified_1.extractType)(P)], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return WeightHook.reified;
    }
    static phantom(P) {
        return (0, reified_1.phantom)(WeightHook.reified(P));
    }
    static get p() {
        return WeightHook.phantom;
    }
    static get bcs() {
        return bcs_1.bcs.struct("WeightHook", {
            id: structs_3.UID.bcs,
            validator_addresses_and_weights: structs_4.VecMap.bcs(bcs_1.bcs
                .bytes(32)
                .transform({
                input: (val) => (0, utils_1.fromHEX)(val),
                output: (val) => (0, utils_1.toHEX)(val),
            }), bcs_1.bcs.u64()),
            total_weight: bcs_1.bcs.u64(),
            admin_cap: structs_5.AdminCap.bcs,
            version: structs_6.Version.bcs,
            extra_fields: structs_2.Bag.bcs,
        });
    }
    static fromFields(typeArg, fields) {
        return WeightHook.reified(typeArg).new({
            id: (0, reified_1.decodeFromFields)(structs_3.UID.reified(), fields.id),
            validatorAddressesAndWeights: (0, reified_1.decodeFromFields)(structs_4.VecMap.reified("address", "u64"), fields.validator_addresses_and_weights),
            totalWeight: (0, reified_1.decodeFromFields)("u64", fields.total_weight),
            adminCap: (0, reified_1.decodeFromFields)(structs_5.AdminCap.reified(typeArg), fields.admin_cap),
            version: (0, reified_1.decodeFromFields)(structs_6.Version.reified(), fields.version),
            extraFields: (0, reified_1.decodeFromFields)(structs_2.Bag.reified(), fields.extra_fields),
        });
    }
    static fromFieldsWithTypes(typeArg, item) {
        if (!isWeightHook(item.type)) {
            throw new Error("not a WeightHook type");
        }
        (0, reified_1.assertFieldsWithTypesArgsMatch)(item, [typeArg]);
        return WeightHook.reified(typeArg).new({
            id: (0, reified_1.decodeFromFieldsWithTypes)(structs_3.UID.reified(), item.fields.id),
            validatorAddressesAndWeights: (0, reified_1.decodeFromFieldsWithTypes)(structs_4.VecMap.reified("address", "u64"), item.fields.validator_addresses_and_weights),
            totalWeight: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.total_weight),
            adminCap: (0, reified_1.decodeFromFieldsWithTypes)(structs_5.AdminCap.reified(typeArg), item.fields.admin_cap),
            version: (0, reified_1.decodeFromFieldsWithTypes)(structs_6.Version.reified(), item.fields.version),
            extraFields: (0, reified_1.decodeFromFieldsWithTypes)(structs_2.Bag.reified(), item.fields.extra_fields),
        });
    }
    static fromBcs(typeArg, data) {
        return WeightHook.fromFields(typeArg, WeightHook.bcs.parse(data));
    }
    toJSONField() {
        return {
            id: this.id,
            validatorAddressesAndWeights: this.validatorAddressesAndWeights.toJSONField(),
            totalWeight: this.totalWeight.toString(),
            adminCap: this.adminCap.toJSONField(),
            version: this.version.toJSONField(),
            extraFields: this.extraFields.toJSONField(),
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(typeArg, field) {
        return WeightHook.reified(typeArg).new({
            id: (0, reified_1.decodeFromJSONField)(structs_3.UID.reified(), field.id),
            validatorAddressesAndWeights: (0, reified_1.decodeFromJSONField)(structs_4.VecMap.reified("address", "u64"), field.validatorAddressesAndWeights),
            totalWeight: (0, reified_1.decodeFromJSONField)("u64", field.totalWeight),
            adminCap: (0, reified_1.decodeFromJSONField)(structs_5.AdminCap.reified(typeArg), field.adminCap),
            version: (0, reified_1.decodeFromJSONField)(structs_6.Version.reified(), field.version),
            extraFields: (0, reified_1.decodeFromJSONField)(structs_2.Bag.reified(), field.extraFields),
        });
    }
    static fromJSON(typeArg, json) {
        if (json.$typeName !== WeightHook.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        (0, reified_1.assertReifiedTypeArgsMatch)((0, util_1.composeSuiType)(WeightHook.$typeName, (0, reified_1.extractType)(typeArg)), json.$typeArgs, [typeArg]);
        return WeightHook.fromJSONField(typeArg, json);
    }
    static fromSuiParsedData(typeArg, content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isWeightHook(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a WeightHook object`);
        }
        return WeightHook.fromFieldsWithTypes(typeArg, content);
    }
    static fromSuiObjectData(typeArg, data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" || !isWeightHook(data.bcs.type)) {
                throw new Error(`object at is not a WeightHook object`);
            }
            const gotTypeArgs = (0, util_1.parseTypeName)(data.bcs.type).typeArgs;
            if (gotTypeArgs.length !== 1) {
                throw new Error(`type argument mismatch: expected 1 type argument but got '${gotTypeArgs.length}'`);
            }
            const gotTypeArg = (0, util_1.compressSuiType)(gotTypeArgs[0]);
            const expectedTypeArg = (0, util_1.compressSuiType)((0, reified_1.extractType)(typeArg));
            if (gotTypeArg !== (0, util_1.compressSuiType)((0, reified_1.extractType)(typeArg))) {
                throw new Error(`type argument mismatch: expected '${expectedTypeArg}' but got '${gotTypeArg}'`);
            }
            return WeightHook.fromBcs(typeArg, (0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return WeightHook.fromSuiParsedData(typeArg, data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, typeArg, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching WeightHook object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isWeightHook(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a WeightHook object`);
            }
            return WeightHook.fromSuiObjectData(typeArg, res.data);
        });
    }
}
exports.WeightHook = WeightHook;
WeightHook.$typeName = `${index_1.PKG_V1}::weight::WeightHook`;
WeightHook.$numTypeParams = 1;
WeightHook.$isPhantom = [true];
/* ============================== WeightHookAdminCap =============================== */
function isWeightHookAdminCap(type) {
    type = (0, util_1.compressSuiType)(type);
    return type.startsWith(`${index_1.PKG_V1}::weight::WeightHookAdminCap` + "<");
}
class WeightHookAdminCap {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = WeightHookAdminCap.$typeName;
        this.$isPhantom = WeightHookAdminCap.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(WeightHookAdminCap.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.id = fields.id;
    }
    static reified(P) {
        return {
            typeName: WeightHookAdminCap.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(WeightHookAdminCap.$typeName, ...[(0, reified_1.extractType)(P)]),
            typeArgs: [(0, reified_1.extractType)(P)],
            isPhantom: WeightHookAdminCap.$isPhantom,
            reifiedTypeArgs: [P],
            fromFields: (fields) => WeightHookAdminCap.fromFields(P, fields),
            fromFieldsWithTypes: (item) => WeightHookAdminCap.fromFieldsWithTypes(P, item),
            fromBcs: (data) => WeightHookAdminCap.fromBcs(P, data),
            bcs: WeightHookAdminCap.bcs,
            fromJSONField: (field) => WeightHookAdminCap.fromJSONField(P, field),
            fromJSON: (json) => WeightHookAdminCap.fromJSON(P, json),
            fromSuiParsedData: (content) => WeightHookAdminCap.fromSuiParsedData(P, content),
            fromSuiObjectData: (content) => WeightHookAdminCap.fromSuiObjectData(P, content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return WeightHookAdminCap.fetch(client, P, id); }),
            new: (fields) => {
                return new WeightHookAdminCap([(0, reified_1.extractType)(P)], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return WeightHookAdminCap.reified;
    }
    static phantom(P) {
        return (0, reified_1.phantom)(WeightHookAdminCap.reified(P));
    }
    static get p() {
        return WeightHookAdminCap.phantom;
    }
    static get bcs() {
        return bcs_1.bcs.struct("WeightHookAdminCap", {
            id: structs_3.UID.bcs,
        });
    }
    static fromFields(typeArg, fields) {
        return WeightHookAdminCap.reified(typeArg).new({
            id: (0, reified_1.decodeFromFields)(structs_3.UID.reified(), fields.id),
        });
    }
    static fromFieldsWithTypes(typeArg, item) {
        if (!isWeightHookAdminCap(item.type)) {
            throw new Error("not a WeightHookAdminCap type");
        }
        (0, reified_1.assertFieldsWithTypesArgsMatch)(item, [typeArg]);
        return WeightHookAdminCap.reified(typeArg).new({
            id: (0, reified_1.decodeFromFieldsWithTypes)(structs_3.UID.reified(), item.fields.id),
        });
    }
    static fromBcs(typeArg, data) {
        return WeightHookAdminCap.fromFields(typeArg, WeightHookAdminCap.bcs.parse(data));
    }
    toJSONField() {
        return {
            id: this.id,
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(typeArg, field) {
        return WeightHookAdminCap.reified(typeArg).new({
            id: (0, reified_1.decodeFromJSONField)(structs_3.UID.reified(), field.id),
        });
    }
    static fromJSON(typeArg, json) {
        if (json.$typeName !== WeightHookAdminCap.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        (0, reified_1.assertReifiedTypeArgsMatch)((0, util_1.composeSuiType)(WeightHookAdminCap.$typeName, (0, reified_1.extractType)(typeArg)), json.$typeArgs, [typeArg]);
        return WeightHookAdminCap.fromJSONField(typeArg, json);
    }
    static fromSuiParsedData(typeArg, content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isWeightHookAdminCap(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a WeightHookAdminCap object`);
        }
        return WeightHookAdminCap.fromFieldsWithTypes(typeArg, content);
    }
    static fromSuiObjectData(typeArg, data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" ||
                !isWeightHookAdminCap(data.bcs.type)) {
                throw new Error(`object at is not a WeightHookAdminCap object`);
            }
            const gotTypeArgs = (0, util_1.parseTypeName)(data.bcs.type).typeArgs;
            if (gotTypeArgs.length !== 1) {
                throw new Error(`type argument mismatch: expected 1 type argument but got '${gotTypeArgs.length}'`);
            }
            const gotTypeArg = (0, util_1.compressSuiType)(gotTypeArgs[0]);
            const expectedTypeArg = (0, util_1.compressSuiType)((0, reified_1.extractType)(typeArg));
            if (gotTypeArg !== (0, util_1.compressSuiType)((0, reified_1.extractType)(typeArg))) {
                throw new Error(`type argument mismatch: expected '${expectedTypeArg}' but got '${gotTypeArg}'`);
            }
            return WeightHookAdminCap.fromBcs(typeArg, (0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return WeightHookAdminCap.fromSuiParsedData(typeArg, data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, typeArg, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching WeightHookAdminCap object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isWeightHookAdminCap(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a WeightHookAdminCap object`);
            }
            return WeightHookAdminCap.fromSuiObjectData(typeArg, res.data);
        });
    }
}
exports.WeightHookAdminCap = WeightHookAdminCap;
WeightHookAdminCap.$typeName = `${index_1.PKG_V1}::weight::WeightHookAdminCap`;
WeightHookAdminCap.$numTypeParams = 1;
WeightHookAdminCap.$isPhantom = [true];
