// @suilend/springsui-sdk/_generated/liquid_staking/version/structs.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Version = void 0;
exports.isVersion = isVersion;
const reified_1 = require("../../_framework/reified");
const util_1 = require("../../_framework/util");
const index_1 = require("../index");
const bcs_1 = require("@mysten/sui/bcs");
const utils_1 = require("@mysten/sui/utils");
/* ============================== Version =============================== */
function isVersion(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V1}::version::Version`;
}
class Version {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = Version.$typeName;
        this.$isPhantom = Version.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(Version.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.pos0 = fields.pos0;
    }
    static reified() {
        return {
            typeName: Version.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(Version.$typeName, ...[]),
            typeArgs: [],
            isPhantom: Version.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => Version.fromFields(fields),
            fromFieldsWithTypes: (item) => Version.fromFieldsWithTypes(item),
            fromBcs: (data) => Version.fromBcs(data),
            bcs: Version.bcs,
            fromJSONField: (field) => Version.fromJSONField(field),
            fromJSON: (json) => Version.fromJSON(json),
            fromSuiParsedData: (content) => Version.fromSuiParsedData(content),
            fromSuiObjectData: (content) => Version.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return Version.fetch(client, id); }),
            new: (fields) => {
                return new Version([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return Version.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(Version.reified());
    }
    static get p() {
        return Version.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("Version", {
            pos0: bcs_1.bcs.u16(),
        });
    }
    static fromFields(fields) {
        return Version.reified().new({
            pos0: (0, reified_1.decodeFromFields)("u16", fields.pos0),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isVersion(item.type)) {
            throw new Error("not a Version type");
        }
        return Version.reified().new({
            pos0: (0, reified_1.decodeFromFieldsWithTypes)("u16", item.fields.pos0),
        });
    }
    static fromBcs(data) {
        return Version.fromFields(Version.bcs.parse(data));
    }
    toJSONField() {
        return {
            pos0: this.pos0,
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return Version.reified().new({
            pos0: (0, reified_1.decodeFromJSONField)("u16", field.pos0),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== Version.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return Version.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isVersion(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a Version object`);
        }
        return Version.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" || !isVersion(data.bcs.type)) {
                throw new Error(`object at is not a Version object`);
            }
            return Version.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return Version.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching Version object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isVersion(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a Version object`);
            }
            return Version.fromSuiObjectData(res.data);
        });
    }
}
exports.Version = Version;
Version.$typeName = `${index_1.PKG_V1}::version::Version`;
Version.$numTypeParams = 0;
Version.$isPhantom = [];
