// @suilend/springsui-sdk/_generated/liquid_staking/fees/functions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.destroy = destroy;
exports.calculateMintFee = calculateMintFee;
exports.calculateRedeemFee = calculateRedeemFee;
exports.customRedeemFeeBps = customRedeemFeeBps;
exports.newBuilder = newBuilder;
exports.redeemFeeBps = redeemFeeBps;
exports.setCustomRedeemFeeBps = setCustomRedeemFeeBps;
exports.setRedeemFeeBps = setRedeemFeeBps;
exports.setSpreadFeeBps = setSpreadFeeBps;
exports.setStakedSuiMintFeeBps = setStakedSuiMintFeeBps;
exports.setStakedSuiRedeemFeeBps = setStakedSuiRedeemFeeBps;
exports.setSuiMintFeeBps = setSuiMintFeeBps;
exports.spreadFeeBps = spreadFeeBps;
exports.stakedSuiMintFeeBps = stakedSuiMintFeeBps;
exports.stakedSuiRedeemFeeBps = stakedSuiRedeemFeeBps;
exports.suiMintFeeBps = suiMintFeeBps;
exports.toFeeConfig = toFeeConfig;
exports.validateFees = validateFees;
const __1 = require("..");
const util_1 = require("../../_framework/util");
function destroy(tx, fees) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::destroy`,
        arguments: [(0, util_1.obj)(tx, fees)],
    });
}
function calculateMintFee(tx, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::calculate_mint_fee`,
        arguments: [(0, util_1.obj)(tx, args.self), (0, util_1.pure)(tx, args.suiAmount, `u64`)],
    });
}
function calculateRedeemFee(tx, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::calculate_redeem_fee`,
        arguments: [(0, util_1.obj)(tx, args.self), (0, util_1.pure)(tx, args.suiAmount, `u64`)],
    });
}
function customRedeemFeeBps(tx, fees) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::custom_redeem_fee_bps`,
        arguments: [(0, util_1.obj)(tx, fees)],
    });
}
function newBuilder(tx) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::new_builder`,
        arguments: [],
    });
}
function redeemFeeBps(tx, fees) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::redeem_fee_bps`,
        arguments: [(0, util_1.obj)(tx, fees)],
    });
}
function setCustomRedeemFeeBps(tx, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::set_custom_redeem_fee_bps`,
        arguments: [(0, util_1.obj)(tx, args.self), (0, util_1.pure)(tx, args.fee, `u64`)],
    });
}
function setRedeemFeeBps(tx, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::set_redeem_fee_bps`,
        arguments: [(0, util_1.obj)(tx, args.self), (0, util_1.pure)(tx, args.fee, `u64`)],
    });
}
function setSpreadFeeBps(tx, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::set_spread_fee_bps`,
        arguments: [(0, util_1.obj)(tx, args.self), (0, util_1.pure)(tx, args.fee, `u64`)],
    });
}
function setStakedSuiMintFeeBps(tx, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::set_staked_sui_mint_fee_bps`,
        arguments: [(0, util_1.obj)(tx, args.self), (0, util_1.pure)(tx, args.fee, `u64`)],
    });
}
function setStakedSuiRedeemFeeBps(tx, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::set_staked_sui_redeem_fee_bps`,
        arguments: [(0, util_1.obj)(tx, args.self), (0, util_1.pure)(tx, args.fee, `u64`)],
    });
}
function setSuiMintFeeBps(tx, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::set_sui_mint_fee_bps`,
        arguments: [(0, util_1.obj)(tx, args.self), (0, util_1.pure)(tx, args.fee, `u64`)],
    });
}
function spreadFeeBps(tx, fees) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::spread_fee_bps`,
        arguments: [(0, util_1.obj)(tx, fees)],
    });
}
function stakedSuiMintFeeBps(tx, fees) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::staked_sui_mint_fee_bps`,
        arguments: [(0, util_1.obj)(tx, fees)],
    });
}
function stakedSuiRedeemFeeBps(tx, fees) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::staked_sui_redeem_fee_bps`,
        arguments: [(0, util_1.obj)(tx, fees)],
    });
}
function suiMintFeeBps(tx, fees) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::sui_mint_fee_bps`,
        arguments: [(0, util_1.obj)(tx, fees)],
    });
}
function toFeeConfig(tx, builder) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::to_fee_config`,
        arguments: [(0, util_1.obj)(tx, builder)],
    });
}
function validateFees(tx, fees) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::fees::validate_fees`,
        arguments: [(0, util_1.obj)(tx, fees)],
    });
}
