// @suilend/springsui-sdk/_generated/liquid_staking/fees/structs.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeeConfigBuilder = exports.FeeConfig = void 0;
exports.isFeeConfig = isFeeConfig;
exports.isFeeConfigBuilder = isFeeConfigBuilder;
const structs_1 = require("../../_dependencies/source/0x2/bag/structs");
const reified_1 = require("../../_framework/reified");
const util_1 = require("../../_framework/util");
const index_1 = require("../index");
const bcs_1 = require("@mysten/sui/bcs");
const utils_1 = require("@mysten/sui/utils");
/* ============================== FeeConfig =============================== */
function isFeeConfig(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V1}::fees::FeeConfig`;
}
class FeeConfig {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = FeeConfig.$typeName;
        this.$isPhantom = FeeConfig.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(FeeConfig.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.suiMintFeeBps = fields.suiMintFeeBps;
        this.stakedSuiMintFeeBps = fields.stakedSuiMintFeeBps;
        this.redeemFeeBps = fields.redeemFeeBps;
        this.stakedSuiRedeemFeeBps = fields.stakedSuiRedeemFeeBps;
        this.spreadFeeBps = fields.spreadFeeBps;
        this.customRedeemFeeBps = fields.customRedeemFeeBps;
        this.extraFields = fields.extraFields;
    }
    static reified() {
        return {
            typeName: FeeConfig.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(FeeConfig.$typeName, ...[]),
            typeArgs: [],
            isPhantom: FeeConfig.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => FeeConfig.fromFields(fields),
            fromFieldsWithTypes: (item) => FeeConfig.fromFieldsWithTypes(item),
            fromBcs: (data) => FeeConfig.fromBcs(data),
            bcs: FeeConfig.bcs,
            fromJSONField: (field) => FeeConfig.fromJSONField(field),
            fromJSON: (json) => FeeConfig.fromJSON(json),
            fromSuiParsedData: (content) => FeeConfig.fromSuiParsedData(content),
            fromSuiObjectData: (content) => FeeConfig.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return FeeConfig.fetch(client, id); }),
            new: (fields) => {
                return new FeeConfig([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return FeeConfig.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(FeeConfig.reified());
    }
    static get p() {
        return FeeConfig.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("FeeConfig", {
            sui_mint_fee_bps: bcs_1.bcs.u64(),
            staked_sui_mint_fee_bps: bcs_1.bcs.u64(),
            redeem_fee_bps: bcs_1.bcs.u64(),
            staked_sui_redeem_fee_bps: bcs_1.bcs.u64(),
            spread_fee_bps: bcs_1.bcs.u64(),
            custom_redeem_fee_bps: bcs_1.bcs.u64(),
            extra_fields: structs_1.Bag.bcs,
        });
    }
    static fromFields(fields) {
        return FeeConfig.reified().new({
            suiMintFeeBps: (0, reified_1.decodeFromFields)("u64", fields.sui_mint_fee_bps),
            stakedSuiMintFeeBps: (0, reified_1.decodeFromFields)("u64", fields.staked_sui_mint_fee_bps),
            redeemFeeBps: (0, reified_1.decodeFromFields)("u64", fields.redeem_fee_bps),
            stakedSuiRedeemFeeBps: (0, reified_1.decodeFromFields)("u64", fields.staked_sui_redeem_fee_bps),
            spreadFeeBps: (0, reified_1.decodeFromFields)("u64", fields.spread_fee_bps),
            customRedeemFeeBps: (0, reified_1.decodeFromFields)("u64", fields.custom_redeem_fee_bps),
            extraFields: (0, reified_1.decodeFromFields)(structs_1.Bag.reified(), fields.extra_fields),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isFeeConfig(item.type)) {
            throw new Error("not a FeeConfig type");
        }
        return FeeConfig.reified().new({
            suiMintFeeBps: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.sui_mint_fee_bps),
            stakedSuiMintFeeBps: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.staked_sui_mint_fee_bps),
            redeemFeeBps: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.redeem_fee_bps),
            stakedSuiRedeemFeeBps: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.staked_sui_redeem_fee_bps),
            spreadFeeBps: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.spread_fee_bps),
            customRedeemFeeBps: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.custom_redeem_fee_bps),
            extraFields: (0, reified_1.decodeFromFieldsWithTypes)(structs_1.Bag.reified(), item.fields.extra_fields),
        });
    }
    static fromBcs(data) {
        return FeeConfig.fromFields(FeeConfig.bcs.parse(data));
    }
    toJSONField() {
        return {
            suiMintFeeBps: this.suiMintFeeBps.toString(),
            stakedSuiMintFeeBps: this.stakedSuiMintFeeBps.toString(),
            redeemFeeBps: this.redeemFeeBps.toString(),
            stakedSuiRedeemFeeBps: this.stakedSuiRedeemFeeBps.toString(),
            spreadFeeBps: this.spreadFeeBps.toString(),
            customRedeemFeeBps: this.customRedeemFeeBps.toString(),
            extraFields: this.extraFields.toJSONField(),
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return FeeConfig.reified().new({
            suiMintFeeBps: (0, reified_1.decodeFromJSONField)("u64", field.suiMintFeeBps),
            stakedSuiMintFeeBps: (0, reified_1.decodeFromJSONField)("u64", field.stakedSuiMintFeeBps),
            redeemFeeBps: (0, reified_1.decodeFromJSONField)("u64", field.redeemFeeBps),
            stakedSuiRedeemFeeBps: (0, reified_1.decodeFromJSONField)("u64", field.stakedSuiRedeemFeeBps),
            spreadFeeBps: (0, reified_1.decodeFromJSONField)("u64", field.spreadFeeBps),
            customRedeemFeeBps: (0, reified_1.decodeFromJSONField)("u64", field.customRedeemFeeBps),
            extraFields: (0, reified_1.decodeFromJSONField)(structs_1.Bag.reified(), field.extraFields),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== FeeConfig.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return FeeConfig.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isFeeConfig(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a FeeConfig object`);
        }
        return FeeConfig.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" || !isFeeConfig(data.bcs.type)) {
                throw new Error(`object at is not a FeeConfig object`);
            }
            return FeeConfig.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return FeeConfig.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching FeeConfig object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isFeeConfig(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a FeeConfig object`);
            }
            return FeeConfig.fromSuiObjectData(res.data);
        });
    }
}
exports.FeeConfig = FeeConfig;
FeeConfig.$typeName = `${index_1.PKG_V1}::fees::FeeConfig`;
FeeConfig.$numTypeParams = 0;
FeeConfig.$isPhantom = [];
/* ============================== FeeConfigBuilder =============================== */
function isFeeConfigBuilder(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V1}::fees::FeeConfigBuilder`;
}
class FeeConfigBuilder {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = FeeConfigBuilder.$typeName;
        this.$isPhantom = FeeConfigBuilder.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(FeeConfigBuilder.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.fields = fields.fields;
    }
    static reified() {
        return {
            typeName: FeeConfigBuilder.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(FeeConfigBuilder.$typeName, ...[]),
            typeArgs: [],
            isPhantom: FeeConfigBuilder.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => FeeConfigBuilder.fromFields(fields),
            fromFieldsWithTypes: (item) => FeeConfigBuilder.fromFieldsWithTypes(item),
            fromBcs: (data) => FeeConfigBuilder.fromBcs(data),
            bcs: FeeConfigBuilder.bcs,
            fromJSONField: (field) => FeeConfigBuilder.fromJSONField(field),
            fromJSON: (json) => FeeConfigBuilder.fromJSON(json),
            fromSuiParsedData: (content) => FeeConfigBuilder.fromSuiParsedData(content),
            fromSuiObjectData: (content) => FeeConfigBuilder.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return FeeConfigBuilder.fetch(client, id); }),
            new: (fields) => {
                return new FeeConfigBuilder([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return FeeConfigBuilder.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(FeeConfigBuilder.reified());
    }
    static get p() {
        return FeeConfigBuilder.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("FeeConfigBuilder", {
            fields: structs_1.Bag.bcs,
        });
    }
    static fromFields(fields) {
        return FeeConfigBuilder.reified().new({
            fields: (0, reified_1.decodeFromFields)(structs_1.Bag.reified(), fields.fields),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isFeeConfigBuilder(item.type)) {
            throw new Error("not a FeeConfigBuilder type");
        }
        return FeeConfigBuilder.reified().new({
            fields: (0, reified_1.decodeFromFieldsWithTypes)(structs_1.Bag.reified(), item.fields.fields),
        });
    }
    static fromBcs(data) {
        return FeeConfigBuilder.fromFields(FeeConfigBuilder.bcs.parse(data));
    }
    toJSONField() {
        return {
            fields: this.fields.toJSONField(),
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return FeeConfigBuilder.reified().new({
            fields: (0, reified_1.decodeFromJSONField)(structs_1.Bag.reified(), field.fields),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== FeeConfigBuilder.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return FeeConfigBuilder.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isFeeConfigBuilder(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a FeeConfigBuilder object`);
        }
        return FeeConfigBuilder.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" ||
                !isFeeConfigBuilder(data.bcs.type)) {
                throw new Error(`object at is not a FeeConfigBuilder object`);
            }
            return FeeConfigBuilder.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return FeeConfigBuilder.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching FeeConfigBuilder object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isFeeConfigBuilder(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a FeeConfigBuilder object`);
            }
            return FeeConfigBuilder.fromSuiObjectData(res.data);
        });
    }
}
exports.FeeConfigBuilder = FeeConfigBuilder;
FeeConfigBuilder.$typeName = `${index_1.PKG_V1}::fees::FeeConfigBuilder`;
FeeConfigBuilder.$numTypeParams = 0;
FeeConfigBuilder.$isPhantom = [];
