// @suilend/springsui-sdk/_generated/liquid_staking/liquid-staking/functions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mint = mint;
exports.storage = storage;
exports.fees = fees;
exports.changeValidatorPriority = changeValidatorPriority;
exports.refresh = refresh;
exports.totalSuiSupply = totalSuiSupply;
exports.collectFees = collectFees;
exports.createLst = createLst;
exports.feeConfig = feeConfig;
exports.createLstWithStake = createLstWithStake;
exports.createLstWithStorage = createLstWithStorage;
exports.decreaseValidatorStake = decreaseValidatorStake;
exports.increaseValidatorStake = increaseValidatorStake;
exports.init = init;
exports.lstAmountToSuiAmount = lstAmountToSuiAmount;
exports.redeem = redeem;
exports.suiAmountToLstAmount = suiAmountToLstAmount;
exports.totalLstSupply = totalLstSupply;
exports.updateFees = updateFees;
const __1 = require("..");
const structs_1 = require("../../_dependencies/source/0x3/staking-pool/structs");
const util_1 = require("../../_framework/util");
function mint(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::mint`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.self),
            (0, util_1.obj)(tx, args.systemState),
            (0, util_1.obj)(tx, args.sui),
        ],
    });
}
function storage(tx, typeArg, self) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::storage`,
        typeArguments: [typeArg],
        arguments: [(0, util_1.obj)(tx, self)],
    });
}
function fees(tx, typeArg, self) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::fees`,
        typeArguments: [typeArg],
        arguments: [(0, util_1.obj)(tx, self)],
    });
}
function changeValidatorPriority(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::change_validator_priority`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.self),
            (0, util_1.obj)(tx, args.adminCap),
            (0, util_1.pure)(tx, args.validatorIndex, `u64`),
            (0, util_1.pure)(tx, args.newValidatorIndex, `u64`),
        ],
    });
}
function refresh(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::refresh`,
        typeArguments: [typeArg],
        arguments: [(0, util_1.obj)(tx, args.self), (0, util_1.obj)(tx, args.systemState)],
    });
}
function totalSuiSupply(tx, typeArg, self) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::total_sui_supply`,
        typeArguments: [typeArg],
        arguments: [(0, util_1.obj)(tx, self)],
    });
}
function collectFees(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::collect_fees`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.self),
            (0, util_1.obj)(tx, args.systemState),
            (0, util_1.obj)(tx, args.adminCap),
        ],
    });
}
function createLst(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::create_lst`,
        typeArguments: [typeArg],
        arguments: [(0, util_1.obj)(tx, args.feeConfig), (0, util_1.obj)(tx, args.lstTreasuryCap)],
    });
}
function feeConfig(tx, typeArg, self) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::fee_config`,
        typeArguments: [typeArg],
        arguments: [(0, util_1.obj)(tx, self)],
    });
}
function createLstWithStake(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::create_lst_with_stake`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.systemState),
            (0, util_1.obj)(tx, args.feeConfig),
            (0, util_1.obj)(tx, args.lstTreasuryCap),
            (0, util_1.vector)(tx, `${structs_1.FungibleStakedSui.$typeName}`, args.fungibleStakedSuis),
            (0, util_1.obj)(tx, args.sui),
        ],
    });
}
function createLstWithStorage(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::create_lst_with_storage`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.feeConfig),
            (0, util_1.obj)(tx, args.lstTreasuryCap),
            (0, util_1.obj)(tx, args.storage),
        ],
    });
}
function decreaseValidatorStake(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::decrease_validator_stake`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.self),
            (0, util_1.obj)(tx, args.adminCap),
            (0, util_1.obj)(tx, args.systemState),
            (0, util_1.pure)(tx, args.validatorAddress, `address`),
            (0, util_1.pure)(tx, args.targetUnstakeSuiAmount, `u64`),
        ],
    });
}
function increaseValidatorStake(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::increase_validator_stake`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.self),
            (0, util_1.obj)(tx, args.adminCap),
            (0, util_1.obj)(tx, args.systemState),
            (0, util_1.pure)(tx, args.validatorAddress, `address`),
            (0, util_1.pure)(tx, args.suiAmount, `u64`),
        ],
    });
}
function init(tx, otw) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::init`,
        arguments: [(0, util_1.obj)(tx, otw)],
    });
}
function lstAmountToSuiAmount(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::lst_amount_to_sui_amount`,
        typeArguments: [typeArg],
        arguments: [(0, util_1.obj)(tx, args.self), (0, util_1.pure)(tx, args.lstAmount, `u64`)],
    });
}
function redeem(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::redeem`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.self),
            (0, util_1.obj)(tx, args.lst),
            (0, util_1.obj)(tx, args.systemState),
        ],
    });
}
function suiAmountToLstAmount(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::sui_amount_to_lst_amount`,
        typeArguments: [typeArg],
        arguments: [(0, util_1.obj)(tx, args.self), (0, util_1.pure)(tx, args.suiAmount, `u64`)],
    });
}
function totalLstSupply(tx, typeArg, self) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::total_lst_supply`,
        typeArguments: [typeArg],
        arguments: [(0, util_1.obj)(tx, self)],
    });
}
function updateFees(tx, typeArg, args) {
    return tx.moveCall({
        target: `${__1.PUBLISHED_AT}::liquid_staking::update_fees`,
        typeArguments: [typeArg],
        arguments: [
            (0, util_1.obj)(tx, args.self),
            (0, util_1.obj)(tx, args.adminCap),
            (0, util_1.obj)(tx, args.feeConfig),
        ],
    });
}
