// @suilend/springsui-sdk/_generated/liquid_staking/liquid-staking/structs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedeemEvent = exports.MintEvent = exports.LiquidStakingInfo = exports.LIQUID_STAKING = exports.IncreaseValidatorStakeEvent = exports.EpochChangedEvent = exports.DecreaseValidatorStakeEvent = exports.CreateEvent = exports.CollectFeesEvent = exports.AdminCap = void 0;
exports.isAdminCap = isAdminCap;
exports.isCollectFeesEvent = isCollectFeesEvent;
exports.isCreateEvent = isCreateEvent;
exports.isDecreaseValidatorStakeEvent = isDecreaseValidatorStakeEvent;
exports.isEpochChangedEvent = isEpochChangedEvent;
exports.isIncreaseValidatorStakeEvent = isIncreaseValidatorStakeEvent;
exports.isLIQUID_STAKING = isLIQUID_STAKING;
exports.isLiquidStakingInfo = isLiquidStakingInfo;
exports.isMintEvent = isMintEvent;
exports.isRedeemEvent = isRedeemEvent;
const reified = __importStar(require("../../_framework/reified"));
const structs_1 = require("../../_dependencies/source/0x1/type-name/structs");
const structs_2 = require("../../_dependencies/source/0x2/bag/structs");
const structs_3 = require("../../_dependencies/source/0x2/balance/structs");
const structs_4 = require("../../_dependencies/source/0x2/coin/structs");
const structs_5 = require("../../_dependencies/source/0x2/object/structs");
const structs_6 = require("../../_dependencies/source/0x2/sui/structs");
const reified_1 = require("../../_framework/reified");
const util_1 = require("../../_framework/util");
const structs_7 = require("../cell/structs");
const structs_8 = require("../fees/structs");
const index_1 = require("../index");
const structs_9 = require("../storage/structs");
const structs_10 = require("../version/structs");
const bcs_1 = require("@mysten/sui/bcs");
const utils_1 = require("@mysten/sui/utils");
/* ============================== AdminCap =============================== */
function isAdminCap(type) {
    type = (0, util_1.compressSuiType)(type);
    return type.startsWith(`${index_1.PKG_V1}::liquid_staking::AdminCap` + "<");
}
class AdminCap {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = AdminCap.$typeName;
        this.$isPhantom = AdminCap.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(AdminCap.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.id = fields.id;
    }
    static reified(P) {
        return {
            typeName: AdminCap.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(AdminCap.$typeName, ...[(0, reified_1.extractType)(P)]),
            typeArgs: [(0, reified_1.extractType)(P)],
            isPhantom: AdminCap.$isPhantom,
            reifiedTypeArgs: [P],
            fromFields: (fields) => AdminCap.fromFields(P, fields),
            fromFieldsWithTypes: (item) => AdminCap.fromFieldsWithTypes(P, item),
            fromBcs: (data) => AdminCap.fromBcs(P, data),
            bcs: AdminCap.bcs,
            fromJSONField: (field) => AdminCap.fromJSONField(P, field),
            fromJSON: (json) => AdminCap.fromJSON(P, json),
            fromSuiParsedData: (content) => AdminCap.fromSuiParsedData(P, content),
            fromSuiObjectData: (content) => AdminCap.fromSuiObjectData(P, content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return AdminCap.fetch(client, P, id); }),
            new: (fields) => {
                return new AdminCap([(0, reified_1.extractType)(P)], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return AdminCap.reified;
    }
    static phantom(P) {
        return (0, reified_1.phantom)(AdminCap.reified(P));
    }
    static get p() {
        return AdminCap.phantom;
    }
    static get bcs() {
        return bcs_1.bcs.struct("AdminCap", {
            id: structs_5.UID.bcs,
        });
    }
    static fromFields(typeArg, fields) {
        return AdminCap.reified(typeArg).new({
            id: (0, reified_1.decodeFromFields)(structs_5.UID.reified(), fields.id),
        });
    }
    static fromFieldsWithTypes(typeArg, item) {
        if (!isAdminCap(item.type)) {
            throw new Error("not a AdminCap type");
        }
        (0, reified_1.assertFieldsWithTypesArgsMatch)(item, [typeArg]);
        return AdminCap.reified(typeArg).new({
            id: (0, reified_1.decodeFromFieldsWithTypes)(structs_5.UID.reified(), item.fields.id),
        });
    }
    static fromBcs(typeArg, data) {
        return AdminCap.fromFields(typeArg, AdminCap.bcs.parse(data));
    }
    toJSONField() {
        return {
            id: this.id,
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(typeArg, field) {
        return AdminCap.reified(typeArg).new({
            id: (0, reified_1.decodeFromJSONField)(structs_5.UID.reified(), field.id),
        });
    }
    static fromJSON(typeArg, json) {
        if (json.$typeName !== AdminCap.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        (0, reified_1.assertReifiedTypeArgsMatch)((0, util_1.composeSuiType)(AdminCap.$typeName, (0, reified_1.extractType)(typeArg)), json.$typeArgs, [typeArg]);
        return AdminCap.fromJSONField(typeArg, json);
    }
    static fromSuiParsedData(typeArg, content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isAdminCap(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a AdminCap object`);
        }
        return AdminCap.fromFieldsWithTypes(typeArg, content);
    }
    static fromSuiObjectData(typeArg, data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" || !isAdminCap(data.bcs.type)) {
                throw new Error(`object at is not a AdminCap object`);
            }
            const gotTypeArgs = (0, util_1.parseTypeName)(data.bcs.type).typeArgs;
            if (gotTypeArgs.length !== 1) {
                throw new Error(`type argument mismatch: expected 1 type argument but got '${gotTypeArgs.length}'`);
            }
            const gotTypeArg = (0, util_1.compressSuiType)(gotTypeArgs[0]);
            const expectedTypeArg = (0, util_1.compressSuiType)((0, reified_1.extractType)(typeArg));
            if (gotTypeArg !== (0, util_1.compressSuiType)((0, reified_1.extractType)(typeArg))) {
                throw new Error(`type argument mismatch: expected '${expectedTypeArg}' but got '${gotTypeArg}'`);
            }
            return AdminCap.fromBcs(typeArg, (0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return AdminCap.fromSuiParsedData(typeArg, data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, typeArg, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching AdminCap object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isAdminCap(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a AdminCap object`);
            }
            return AdminCap.fromSuiObjectData(typeArg, res.data);
        });
    }
}
exports.AdminCap = AdminCap;
AdminCap.$typeName = `${index_1.PKG_V1}::liquid_staking::AdminCap`;
AdminCap.$numTypeParams = 1;
AdminCap.$isPhantom = [true];
/* ============================== CollectFeesEvent =============================== */
function isCollectFeesEvent(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V1}::liquid_staking::CollectFeesEvent`;
}
class CollectFeesEvent {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = CollectFeesEvent.$typeName;
        this.$isPhantom = CollectFeesEvent.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(CollectFeesEvent.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.typename = fields.typename;
        this.amount = fields.amount;
    }
    static reified() {
        return {
            typeName: CollectFeesEvent.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(CollectFeesEvent.$typeName, ...[]),
            typeArgs: [],
            isPhantom: CollectFeesEvent.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => CollectFeesEvent.fromFields(fields),
            fromFieldsWithTypes: (item) => CollectFeesEvent.fromFieldsWithTypes(item),
            fromBcs: (data) => CollectFeesEvent.fromBcs(data),
            bcs: CollectFeesEvent.bcs,
            fromJSONField: (field) => CollectFeesEvent.fromJSONField(field),
            fromJSON: (json) => CollectFeesEvent.fromJSON(json),
            fromSuiParsedData: (content) => CollectFeesEvent.fromSuiParsedData(content),
            fromSuiObjectData: (content) => CollectFeesEvent.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return CollectFeesEvent.fetch(client, id); }),
            new: (fields) => {
                return new CollectFeesEvent([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return CollectFeesEvent.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(CollectFeesEvent.reified());
    }
    static get p() {
        return CollectFeesEvent.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("CollectFeesEvent", {
            typename: structs_1.TypeName.bcs,
            amount: bcs_1.bcs.u64(),
        });
    }
    static fromFields(fields) {
        return CollectFeesEvent.reified().new({
            typename: (0, reified_1.decodeFromFields)(structs_1.TypeName.reified(), fields.typename),
            amount: (0, reified_1.decodeFromFields)("u64", fields.amount),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isCollectFeesEvent(item.type)) {
            throw new Error("not a CollectFeesEvent type");
        }
        return CollectFeesEvent.reified().new({
            typename: (0, reified_1.decodeFromFieldsWithTypes)(structs_1.TypeName.reified(), item.fields.typename),
            amount: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.amount),
        });
    }
    static fromBcs(data) {
        return CollectFeesEvent.fromFields(CollectFeesEvent.bcs.parse(data));
    }
    toJSONField() {
        return {
            typename: this.typename.toJSONField(),
            amount: this.amount.toString(),
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return CollectFeesEvent.reified().new({
            typename: (0, reified_1.decodeFromJSONField)(structs_1.TypeName.reified(), field.typename),
            amount: (0, reified_1.decodeFromJSONField)("u64", field.amount),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== CollectFeesEvent.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return CollectFeesEvent.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isCollectFeesEvent(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a CollectFeesEvent object`);
        }
        return CollectFeesEvent.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" ||
                !isCollectFeesEvent(data.bcs.type)) {
                throw new Error(`object at is not a CollectFeesEvent object`);
            }
            return CollectFeesEvent.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return CollectFeesEvent.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching CollectFeesEvent object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isCollectFeesEvent(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a CollectFeesEvent object`);
            }
            return CollectFeesEvent.fromSuiObjectData(res.data);
        });
    }
}
exports.CollectFeesEvent = CollectFeesEvent;
CollectFeesEvent.$typeName = `${index_1.PKG_V1}::liquid_staking::CollectFeesEvent`;
CollectFeesEvent.$numTypeParams = 0;
CollectFeesEvent.$isPhantom = [];
/* ============================== CreateEvent =============================== */
function isCreateEvent(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V1}::liquid_staking::CreateEvent`;
}
class CreateEvent {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = CreateEvent.$typeName;
        this.$isPhantom = CreateEvent.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(CreateEvent.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.typename = fields.typename;
        this.liquidStakingInfoId = fields.liquidStakingInfoId;
    }
    static reified() {
        return {
            typeName: CreateEvent.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(CreateEvent.$typeName, ...[]),
            typeArgs: [],
            isPhantom: CreateEvent.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => CreateEvent.fromFields(fields),
            fromFieldsWithTypes: (item) => CreateEvent.fromFieldsWithTypes(item),
            fromBcs: (data) => CreateEvent.fromBcs(data),
            bcs: CreateEvent.bcs,
            fromJSONField: (field) => CreateEvent.fromJSONField(field),
            fromJSON: (json) => CreateEvent.fromJSON(json),
            fromSuiParsedData: (content) => CreateEvent.fromSuiParsedData(content),
            fromSuiObjectData: (content) => CreateEvent.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return CreateEvent.fetch(client, id); }),
            new: (fields) => {
                return new CreateEvent([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return CreateEvent.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(CreateEvent.reified());
    }
    static get p() {
        return CreateEvent.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("CreateEvent", {
            typename: structs_1.TypeName.bcs,
            liquid_staking_info_id: structs_5.ID.bcs,
        });
    }
    static fromFields(fields) {
        return CreateEvent.reified().new({
            typename: (0, reified_1.decodeFromFields)(structs_1.TypeName.reified(), fields.typename),
            liquidStakingInfoId: (0, reified_1.decodeFromFields)(structs_5.ID.reified(), fields.liquid_staking_info_id),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isCreateEvent(item.type)) {
            throw new Error("not a CreateEvent type");
        }
        return CreateEvent.reified().new({
            typename: (0, reified_1.decodeFromFieldsWithTypes)(structs_1.TypeName.reified(), item.fields.typename),
            liquidStakingInfoId: (0, reified_1.decodeFromFieldsWithTypes)(structs_5.ID.reified(), item.fields.liquid_staking_info_id),
        });
    }
    static fromBcs(data) {
        return CreateEvent.fromFields(CreateEvent.bcs.parse(data));
    }
    toJSONField() {
        return {
            typename: this.typename.toJSONField(),
            liquidStakingInfoId: this.liquidStakingInfoId,
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return CreateEvent.reified().new({
            typename: (0, reified_1.decodeFromJSONField)(structs_1.TypeName.reified(), field.typename),
            liquidStakingInfoId: (0, reified_1.decodeFromJSONField)(structs_5.ID.reified(), field.liquidStakingInfoId),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== CreateEvent.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return CreateEvent.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isCreateEvent(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a CreateEvent object`);
        }
        return CreateEvent.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" || !isCreateEvent(data.bcs.type)) {
                throw new Error(`object at is not a CreateEvent object`);
            }
            return CreateEvent.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return CreateEvent.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching CreateEvent object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isCreateEvent(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a CreateEvent object`);
            }
            return CreateEvent.fromSuiObjectData(res.data);
        });
    }
}
exports.CreateEvent = CreateEvent;
CreateEvent.$typeName = `${index_1.PKG_V1}::liquid_staking::CreateEvent`;
CreateEvent.$numTypeParams = 0;
CreateEvent.$isPhantom = [];
/* ============================== DecreaseValidatorStakeEvent =============================== */
function isDecreaseValidatorStakeEvent(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V1}::liquid_staking::DecreaseValidatorStakeEvent`;
}
class DecreaseValidatorStakeEvent {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = DecreaseValidatorStakeEvent.$typeName;
        this.$isPhantom = DecreaseValidatorStakeEvent.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(DecreaseValidatorStakeEvent.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.typename = fields.typename;
        this.stakingPoolId = fields.stakingPoolId;
        this.amount = fields.amount;
    }
    static reified() {
        return {
            typeName: DecreaseValidatorStakeEvent.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(DecreaseValidatorStakeEvent.$typeName, ...[]),
            typeArgs: [],
            isPhantom: DecreaseValidatorStakeEvent.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => DecreaseValidatorStakeEvent.fromFields(fields),
            fromFieldsWithTypes: (item) => DecreaseValidatorStakeEvent.fromFieldsWithTypes(item),
            fromBcs: (data) => DecreaseValidatorStakeEvent.fromBcs(data),
            bcs: DecreaseValidatorStakeEvent.bcs,
            fromJSONField: (field) => DecreaseValidatorStakeEvent.fromJSONField(field),
            fromJSON: (json) => DecreaseValidatorStakeEvent.fromJSON(json),
            fromSuiParsedData: (content) => DecreaseValidatorStakeEvent.fromSuiParsedData(content),
            fromSuiObjectData: (content) => DecreaseValidatorStakeEvent.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return DecreaseValidatorStakeEvent.fetch(client, id); }),
            new: (fields) => {
                return new DecreaseValidatorStakeEvent([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return DecreaseValidatorStakeEvent.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(DecreaseValidatorStakeEvent.reified());
    }
    static get p() {
        return DecreaseValidatorStakeEvent.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("DecreaseValidatorStakeEvent", {
            typename: structs_1.TypeName.bcs,
            staking_pool_id: structs_5.ID.bcs,
            amount: bcs_1.bcs.u64(),
        });
    }
    static fromFields(fields) {
        return DecreaseValidatorStakeEvent.reified().new({
            typename: (0, reified_1.decodeFromFields)(structs_1.TypeName.reified(), fields.typename),
            stakingPoolId: (0, reified_1.decodeFromFields)(structs_5.ID.reified(), fields.staking_pool_id),
            amount: (0, reified_1.decodeFromFields)("u64", fields.amount),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isDecreaseValidatorStakeEvent(item.type)) {
            throw new Error("not a DecreaseValidatorStakeEvent type");
        }
        return DecreaseValidatorStakeEvent.reified().new({
            typename: (0, reified_1.decodeFromFieldsWithTypes)(structs_1.TypeName.reified(), item.fields.typename),
            stakingPoolId: (0, reified_1.decodeFromFieldsWithTypes)(structs_5.ID.reified(), item.fields.staking_pool_id),
            amount: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.amount),
        });
    }
    static fromBcs(data) {
        return DecreaseValidatorStakeEvent.fromFields(DecreaseValidatorStakeEvent.bcs.parse(data));
    }
    toJSONField() {
        return {
            typename: this.typename.toJSONField(),
            stakingPoolId: this.stakingPoolId,
            amount: this.amount.toString(),
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return DecreaseValidatorStakeEvent.reified().new({
            typename: (0, reified_1.decodeFromJSONField)(structs_1.TypeName.reified(), field.typename),
            stakingPoolId: (0, reified_1.decodeFromJSONField)(structs_5.ID.reified(), field.stakingPoolId),
            amount: (0, reified_1.decodeFromJSONField)("u64", field.amount),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== DecreaseValidatorStakeEvent.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return DecreaseValidatorStakeEvent.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isDecreaseValidatorStakeEvent(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a DecreaseValidatorStakeEvent object`);
        }
        return DecreaseValidatorStakeEvent.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" ||
                !isDecreaseValidatorStakeEvent(data.bcs.type)) {
                throw new Error(`object at is not a DecreaseValidatorStakeEvent object`);
            }
            return DecreaseValidatorStakeEvent.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return DecreaseValidatorStakeEvent.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching DecreaseValidatorStakeEvent object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isDecreaseValidatorStakeEvent(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a DecreaseValidatorStakeEvent object`);
            }
            return DecreaseValidatorStakeEvent.fromSuiObjectData(res.data);
        });
    }
}
exports.DecreaseValidatorStakeEvent = DecreaseValidatorStakeEvent;
DecreaseValidatorStakeEvent.$typeName = `${index_1.PKG_V1}::liquid_staking::DecreaseValidatorStakeEvent`;
DecreaseValidatorStakeEvent.$numTypeParams = 0;
DecreaseValidatorStakeEvent.$isPhantom = [];
/* ============================== EpochChangedEvent =============================== */
function isEpochChangedEvent(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V1}::liquid_staking::EpochChangedEvent`;
}
class EpochChangedEvent {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = EpochChangedEvent.$typeName;
        this.$isPhantom = EpochChangedEvent.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(EpochChangedEvent.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.typename = fields.typename;
        this.oldSuiSupply = fields.oldSuiSupply;
        this.newSuiSupply = fields.newSuiSupply;
        this.lstSupply = fields.lstSupply;
        this.spreadFee = fields.spreadFee;
    }
    static reified() {
        return {
            typeName: EpochChangedEvent.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(EpochChangedEvent.$typeName, ...[]),
            typeArgs: [],
            isPhantom: EpochChangedEvent.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => EpochChangedEvent.fromFields(fields),
            fromFieldsWithTypes: (item) => EpochChangedEvent.fromFieldsWithTypes(item),
            fromBcs: (data) => EpochChangedEvent.fromBcs(data),
            bcs: EpochChangedEvent.bcs,
            fromJSONField: (field) => EpochChangedEvent.fromJSONField(field),
            fromJSON: (json) => EpochChangedEvent.fromJSON(json),
            fromSuiParsedData: (content) => EpochChangedEvent.fromSuiParsedData(content),
            fromSuiObjectData: (content) => EpochChangedEvent.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return EpochChangedEvent.fetch(client, id); }),
            new: (fields) => {
                return new EpochChangedEvent([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return EpochChangedEvent.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(EpochChangedEvent.reified());
    }
    static get p() {
        return EpochChangedEvent.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("EpochChangedEvent", {
            typename: structs_1.TypeName.bcs,
            old_sui_supply: bcs_1.bcs.u64(),
            new_sui_supply: bcs_1.bcs.u64(),
            lst_supply: bcs_1.bcs.u64(),
            spread_fee: bcs_1.bcs.u64(),
        });
    }
    static fromFields(fields) {
        return EpochChangedEvent.reified().new({
            typename: (0, reified_1.decodeFromFields)(structs_1.TypeName.reified(), fields.typename),
            oldSuiSupply: (0, reified_1.decodeFromFields)("u64", fields.old_sui_supply),
            newSuiSupply: (0, reified_1.decodeFromFields)("u64", fields.new_sui_supply),
            lstSupply: (0, reified_1.decodeFromFields)("u64", fields.lst_supply),
            spreadFee: (0, reified_1.decodeFromFields)("u64", fields.spread_fee),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isEpochChangedEvent(item.type)) {
            throw new Error("not a EpochChangedEvent type");
        }
        return EpochChangedEvent.reified().new({
            typename: (0, reified_1.decodeFromFieldsWithTypes)(structs_1.TypeName.reified(), item.fields.typename),
            oldSuiSupply: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.old_sui_supply),
            newSuiSupply: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.new_sui_supply),
            lstSupply: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.lst_supply),
            spreadFee: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.spread_fee),
        });
    }
    static fromBcs(data) {
        return EpochChangedEvent.fromFields(EpochChangedEvent.bcs.parse(data));
    }
    toJSONField() {
        return {
            typename: this.typename.toJSONField(),
            oldSuiSupply: this.oldSuiSupply.toString(),
            newSuiSupply: this.newSuiSupply.toString(),
            lstSupply: this.lstSupply.toString(),
            spreadFee: this.spreadFee.toString(),
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return EpochChangedEvent.reified().new({
            typename: (0, reified_1.decodeFromJSONField)(structs_1.TypeName.reified(), field.typename),
            oldSuiSupply: (0, reified_1.decodeFromJSONField)("u64", field.oldSuiSupply),
            newSuiSupply: (0, reified_1.decodeFromJSONField)("u64", field.newSuiSupply),
            lstSupply: (0, reified_1.decodeFromJSONField)("u64", field.lstSupply),
            spreadFee: (0, reified_1.decodeFromJSONField)("u64", field.spreadFee),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== EpochChangedEvent.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return EpochChangedEvent.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isEpochChangedEvent(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a EpochChangedEvent object`);
        }
        return EpochChangedEvent.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" ||
                !isEpochChangedEvent(data.bcs.type)) {
                throw new Error(`object at is not a EpochChangedEvent object`);
            }
            return EpochChangedEvent.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return EpochChangedEvent.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching EpochChangedEvent object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isEpochChangedEvent(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a EpochChangedEvent object`);
            }
            return EpochChangedEvent.fromSuiObjectData(res.data);
        });
    }
}
exports.EpochChangedEvent = EpochChangedEvent;
EpochChangedEvent.$typeName = `${index_1.PKG_V1}::liquid_staking::EpochChangedEvent`;
EpochChangedEvent.$numTypeParams = 0;
EpochChangedEvent.$isPhantom = [];
/* ============================== IncreaseValidatorStakeEvent =============================== */
function isIncreaseValidatorStakeEvent(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V1}::liquid_staking::IncreaseValidatorStakeEvent`;
}
class IncreaseValidatorStakeEvent {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = IncreaseValidatorStakeEvent.$typeName;
        this.$isPhantom = IncreaseValidatorStakeEvent.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(IncreaseValidatorStakeEvent.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.typename = fields.typename;
        this.stakingPoolId = fields.stakingPoolId;
        this.amount = fields.amount;
    }
    static reified() {
        return {
            typeName: IncreaseValidatorStakeEvent.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(IncreaseValidatorStakeEvent.$typeName, ...[]),
            typeArgs: [],
            isPhantom: IncreaseValidatorStakeEvent.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => IncreaseValidatorStakeEvent.fromFields(fields),
            fromFieldsWithTypes: (item) => IncreaseValidatorStakeEvent.fromFieldsWithTypes(item),
            fromBcs: (data) => IncreaseValidatorStakeEvent.fromBcs(data),
            bcs: IncreaseValidatorStakeEvent.bcs,
            fromJSONField: (field) => IncreaseValidatorStakeEvent.fromJSONField(field),
            fromJSON: (json) => IncreaseValidatorStakeEvent.fromJSON(json),
            fromSuiParsedData: (content) => IncreaseValidatorStakeEvent.fromSuiParsedData(content),
            fromSuiObjectData: (content) => IncreaseValidatorStakeEvent.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return IncreaseValidatorStakeEvent.fetch(client, id); }),
            new: (fields) => {
                return new IncreaseValidatorStakeEvent([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return IncreaseValidatorStakeEvent.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(IncreaseValidatorStakeEvent.reified());
    }
    static get p() {
        return IncreaseValidatorStakeEvent.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("IncreaseValidatorStakeEvent", {
            typename: structs_1.TypeName.bcs,
            staking_pool_id: structs_5.ID.bcs,
            amount: bcs_1.bcs.u64(),
        });
    }
    static fromFields(fields) {
        return IncreaseValidatorStakeEvent.reified().new({
            typename: (0, reified_1.decodeFromFields)(structs_1.TypeName.reified(), fields.typename),
            stakingPoolId: (0, reified_1.decodeFromFields)(structs_5.ID.reified(), fields.staking_pool_id),
            amount: (0, reified_1.decodeFromFields)("u64", fields.amount),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isIncreaseValidatorStakeEvent(item.type)) {
            throw new Error("not a IncreaseValidatorStakeEvent type");
        }
        return IncreaseValidatorStakeEvent.reified().new({
            typename: (0, reified_1.decodeFromFieldsWithTypes)(structs_1.TypeName.reified(), item.fields.typename),
            stakingPoolId: (0, reified_1.decodeFromFieldsWithTypes)(structs_5.ID.reified(), item.fields.staking_pool_id),
            amount: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.amount),
        });
    }
    static fromBcs(data) {
        return IncreaseValidatorStakeEvent.fromFields(IncreaseValidatorStakeEvent.bcs.parse(data));
    }
    toJSONField() {
        return {
            typename: this.typename.toJSONField(),
            stakingPoolId: this.stakingPoolId,
            amount: this.amount.toString(),
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return IncreaseValidatorStakeEvent.reified().new({
            typename: (0, reified_1.decodeFromJSONField)(structs_1.TypeName.reified(), field.typename),
            stakingPoolId: (0, reified_1.decodeFromJSONField)(structs_5.ID.reified(), field.stakingPoolId),
            amount: (0, reified_1.decodeFromJSONField)("u64", field.amount),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== IncreaseValidatorStakeEvent.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return IncreaseValidatorStakeEvent.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isIncreaseValidatorStakeEvent(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a IncreaseValidatorStakeEvent object`);
        }
        return IncreaseValidatorStakeEvent.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" ||
                !isIncreaseValidatorStakeEvent(data.bcs.type)) {
                throw new Error(`object at is not a IncreaseValidatorStakeEvent object`);
            }
            return IncreaseValidatorStakeEvent.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return IncreaseValidatorStakeEvent.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching IncreaseValidatorStakeEvent object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isIncreaseValidatorStakeEvent(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a IncreaseValidatorStakeEvent object`);
            }
            return IncreaseValidatorStakeEvent.fromSuiObjectData(res.data);
        });
    }
}
exports.IncreaseValidatorStakeEvent = IncreaseValidatorStakeEvent;
IncreaseValidatorStakeEvent.$typeName = `${index_1.PKG_V1}::liquid_staking::IncreaseValidatorStakeEvent`;
IncreaseValidatorStakeEvent.$numTypeParams = 0;
IncreaseValidatorStakeEvent.$isPhantom = [];
/* ============================== LIQUID_STAKING =============================== */
function isLIQUID_STAKING(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V1}::liquid_staking::LIQUID_STAKING`;
}
class LIQUID_STAKING {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = LIQUID_STAKING.$typeName;
        this.$isPhantom = LIQUID_STAKING.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(LIQUID_STAKING.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.dummyField = fields.dummyField;
    }
    static reified() {
        return {
            typeName: LIQUID_STAKING.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(LIQUID_STAKING.$typeName, ...[]),
            typeArgs: [],
            isPhantom: LIQUID_STAKING.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => LIQUID_STAKING.fromFields(fields),
            fromFieldsWithTypes: (item) => LIQUID_STAKING.fromFieldsWithTypes(item),
            fromBcs: (data) => LIQUID_STAKING.fromBcs(data),
            bcs: LIQUID_STAKING.bcs,
            fromJSONField: (field) => LIQUID_STAKING.fromJSONField(field),
            fromJSON: (json) => LIQUID_STAKING.fromJSON(json),
            fromSuiParsedData: (content) => LIQUID_STAKING.fromSuiParsedData(content),
            fromSuiObjectData: (content) => LIQUID_STAKING.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return LIQUID_STAKING.fetch(client, id); }),
            new: (fields) => {
                return new LIQUID_STAKING([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return LIQUID_STAKING.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(LIQUID_STAKING.reified());
    }
    static get p() {
        return LIQUID_STAKING.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("LIQUID_STAKING", {
            dummy_field: bcs_1.bcs.bool(),
        });
    }
    static fromFields(fields) {
        return LIQUID_STAKING.reified().new({
            dummyField: (0, reified_1.decodeFromFields)("bool", fields.dummy_field),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isLIQUID_STAKING(item.type)) {
            throw new Error("not a LIQUID_STAKING type");
        }
        return LIQUID_STAKING.reified().new({
            dummyField: (0, reified_1.decodeFromFieldsWithTypes)("bool", item.fields.dummy_field),
        });
    }
    static fromBcs(data) {
        return LIQUID_STAKING.fromFields(LIQUID_STAKING.bcs.parse(data));
    }
    toJSONField() {
        return {
            dummyField: this.dummyField,
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return LIQUID_STAKING.reified().new({
            dummyField: (0, reified_1.decodeFromJSONField)("bool", field.dummyField),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== LIQUID_STAKING.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return LIQUID_STAKING.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isLIQUID_STAKING(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a LIQUID_STAKING object`);
        }
        return LIQUID_STAKING.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" ||
                !isLIQUID_STAKING(data.bcs.type)) {
                throw new Error(`object at is not a LIQUID_STAKING object`);
            }
            return LIQUID_STAKING.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return LIQUID_STAKING.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching LIQUID_STAKING object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isLIQUID_STAKING(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a LIQUID_STAKING object`);
            }
            return LIQUID_STAKING.fromSuiObjectData(res.data);
        });
    }
}
exports.LIQUID_STAKING = LIQUID_STAKING;
LIQUID_STAKING.$typeName = `${index_1.PKG_V1}::liquid_staking::LIQUID_STAKING`;
LIQUID_STAKING.$numTypeParams = 0;
LIQUID_STAKING.$isPhantom = [];
/* ============================== LiquidStakingInfo =============================== */
function isLiquidStakingInfo(type) {
    type = (0, util_1.compressSuiType)(type);
    return type.startsWith(`${index_1.PKG_V1}::liquid_staking::LiquidStakingInfo` + "<");
}
class LiquidStakingInfo {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = LiquidStakingInfo.$typeName;
        this.$isPhantom = LiquidStakingInfo.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(LiquidStakingInfo.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.id = fields.id;
        this.lstTreasuryCap = fields.lstTreasuryCap;
        this.feeConfig = fields.feeConfig;
        this.fees = fields.fees;
        this.accruedSpreadFees = fields.accruedSpreadFees;
        this.storage = fields.storage;
        this.version = fields.version;
        this.extraFields = fields.extraFields;
    }
    static reified(P) {
        return {
            typeName: LiquidStakingInfo.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(LiquidStakingInfo.$typeName, ...[(0, reified_1.extractType)(P)]),
            typeArgs: [(0, reified_1.extractType)(P)],
            isPhantom: LiquidStakingInfo.$isPhantom,
            reifiedTypeArgs: [P],
            fromFields: (fields) => LiquidStakingInfo.fromFields(P, fields),
            fromFieldsWithTypes: (item) => LiquidStakingInfo.fromFieldsWithTypes(P, item),
            fromBcs: (data) => LiquidStakingInfo.fromBcs(P, data),
            bcs: LiquidStakingInfo.bcs,
            fromJSONField: (field) => LiquidStakingInfo.fromJSONField(P, field),
            fromJSON: (json) => LiquidStakingInfo.fromJSON(P, json),
            fromSuiParsedData: (content) => LiquidStakingInfo.fromSuiParsedData(P, content),
            fromSuiObjectData: (content) => LiquidStakingInfo.fromSuiObjectData(P, content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return LiquidStakingInfo.fetch(client, P, id); }),
            new: (fields) => {
                return new LiquidStakingInfo([(0, reified_1.extractType)(P)], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return LiquidStakingInfo.reified;
    }
    static phantom(P) {
        return (0, reified_1.phantom)(LiquidStakingInfo.reified(P));
    }
    static get p() {
        return LiquidStakingInfo.phantom;
    }
    static get bcs() {
        return bcs_1.bcs.struct("LiquidStakingInfo", {
            id: structs_5.UID.bcs,
            lst_treasury_cap: structs_4.TreasuryCap.bcs,
            fee_config: structs_7.Cell.bcs(structs_8.FeeConfig.bcs),
            fees: structs_3.Balance.bcs,
            accrued_spread_fees: bcs_1.bcs.u64(),
            storage: structs_9.Storage.bcs,
            version: structs_10.Version.bcs,
            extra_fields: structs_2.Bag.bcs,
        });
    }
    static fromFields(typeArg, fields) {
        return LiquidStakingInfo.reified(typeArg).new({
            id: (0, reified_1.decodeFromFields)(structs_5.UID.reified(), fields.id),
            lstTreasuryCap: (0, reified_1.decodeFromFields)(structs_4.TreasuryCap.reified(typeArg), fields.lst_treasury_cap),
            feeConfig: (0, reified_1.decodeFromFields)(structs_7.Cell.reified(structs_8.FeeConfig.reified()), fields.fee_config),
            fees: (0, reified_1.decodeFromFields)(structs_3.Balance.reified(reified.phantom(structs_6.SUI.reified())), fields.fees),
            accruedSpreadFees: (0, reified_1.decodeFromFields)("u64", fields.accrued_spread_fees),
            storage: (0, reified_1.decodeFromFields)(structs_9.Storage.reified(), fields.storage),
            version: (0, reified_1.decodeFromFields)(structs_10.Version.reified(), fields.version),
            extraFields: (0, reified_1.decodeFromFields)(structs_2.Bag.reified(), fields.extra_fields),
        });
    }
    static fromFieldsWithTypes(typeArg, item) {
        if (!isLiquidStakingInfo(item.type)) {
            throw new Error("not a LiquidStakingInfo type");
        }
        (0, reified_1.assertFieldsWithTypesArgsMatch)(item, [typeArg]);
        return LiquidStakingInfo.reified(typeArg).new({
            id: (0, reified_1.decodeFromFieldsWithTypes)(structs_5.UID.reified(), item.fields.id),
            lstTreasuryCap: (0, reified_1.decodeFromFieldsWithTypes)(structs_4.TreasuryCap.reified(typeArg), item.fields.lst_treasury_cap),
            feeConfig: (0, reified_1.decodeFromFieldsWithTypes)(structs_7.Cell.reified(structs_8.FeeConfig.reified()), item.fields.fee_config),
            fees: (0, reified_1.decodeFromFieldsWithTypes)(structs_3.Balance.reified(reified.phantom(structs_6.SUI.reified())), item.fields.fees),
            accruedSpreadFees: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.accrued_spread_fees),
            storage: (0, reified_1.decodeFromFieldsWithTypes)(structs_9.Storage.reified(), item.fields.storage),
            version: (0, reified_1.decodeFromFieldsWithTypes)(structs_10.Version.reified(), item.fields.version),
            extraFields: (0, reified_1.decodeFromFieldsWithTypes)(structs_2.Bag.reified(), item.fields.extra_fields),
        });
    }
    static fromBcs(typeArg, data) {
        return LiquidStakingInfo.fromFields(typeArg, LiquidStakingInfo.bcs.parse(data));
    }
    toJSONField() {
        return {
            id: this.id,
            lstTreasuryCap: this.lstTreasuryCap.toJSONField(),
            feeConfig: this.feeConfig.toJSONField(),
            fees: this.fees.toJSONField(),
            accruedSpreadFees: this.accruedSpreadFees.toString(),
            storage: this.storage.toJSONField(),
            version: this.version.toJSONField(),
            extraFields: this.extraFields.toJSONField(),
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(typeArg, field) {
        return LiquidStakingInfo.reified(typeArg).new({
            id: (0, reified_1.decodeFromJSONField)(structs_5.UID.reified(), field.id),
            lstTreasuryCap: (0, reified_1.decodeFromJSONField)(structs_4.TreasuryCap.reified(typeArg), field.lstTreasuryCap),
            feeConfig: (0, reified_1.decodeFromJSONField)(structs_7.Cell.reified(structs_8.FeeConfig.reified()), field.feeConfig),
            fees: (0, reified_1.decodeFromJSONField)(structs_3.Balance.reified(reified.phantom(structs_6.SUI.reified())), field.fees),
            accruedSpreadFees: (0, reified_1.decodeFromJSONField)("u64", field.accruedSpreadFees),
            storage: (0, reified_1.decodeFromJSONField)(structs_9.Storage.reified(), field.storage),
            version: (0, reified_1.decodeFromJSONField)(structs_10.Version.reified(), field.version),
            extraFields: (0, reified_1.decodeFromJSONField)(structs_2.Bag.reified(), field.extraFields),
        });
    }
    static fromJSON(typeArg, json) {
        if (json.$typeName !== LiquidStakingInfo.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        (0, reified_1.assertReifiedTypeArgsMatch)((0, util_1.composeSuiType)(LiquidStakingInfo.$typeName, (0, reified_1.extractType)(typeArg)), json.$typeArgs, [typeArg]);
        return LiquidStakingInfo.fromJSONField(typeArg, json);
    }
    static fromSuiParsedData(typeArg, content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isLiquidStakingInfo(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a LiquidStakingInfo object`);
        }
        return LiquidStakingInfo.fromFieldsWithTypes(typeArg, content);
    }
    static fromSuiObjectData(typeArg, data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" ||
                !isLiquidStakingInfo(data.bcs.type)) {
                throw new Error(`object at is not a LiquidStakingInfo object`);
            }
            const gotTypeArgs = (0, util_1.parseTypeName)(data.bcs.type).typeArgs;
            if (gotTypeArgs.length !== 1) {
                throw new Error(`type argument mismatch: expected 1 type argument but got '${gotTypeArgs.length}'`);
            }
            const gotTypeArg = (0, util_1.compressSuiType)(gotTypeArgs[0]);
            const expectedTypeArg = (0, util_1.compressSuiType)((0, reified_1.extractType)(typeArg));
            if (gotTypeArg !== (0, util_1.compressSuiType)((0, reified_1.extractType)(typeArg))) {
                throw new Error(`type argument mismatch: expected '${expectedTypeArg}' but got '${gotTypeArg}'`);
            }
            return LiquidStakingInfo.fromBcs(typeArg, (0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return LiquidStakingInfo.fromSuiParsedData(typeArg, data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, typeArg, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching LiquidStakingInfo object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isLiquidStakingInfo(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a LiquidStakingInfo object`);
            }
            return LiquidStakingInfo.fromSuiObjectData(typeArg, res.data);
        });
    }
}
exports.LiquidStakingInfo = LiquidStakingInfo;
LiquidStakingInfo.$typeName = `${index_1.PKG_V1}::liquid_staking::LiquidStakingInfo`;
LiquidStakingInfo.$numTypeParams = 1;
LiquidStakingInfo.$isPhantom = [true];
/* ============================== MintEvent =============================== */
function isMintEvent(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V1}::liquid_staking::MintEvent`;
}
class MintEvent {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = MintEvent.$typeName;
        this.$isPhantom = MintEvent.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(MintEvent.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.typename = fields.typename;
        this.suiAmountIn = fields.suiAmountIn;
        this.lstAmountOut = fields.lstAmountOut;
        this.feeAmount = fields.feeAmount;
    }
    static reified() {
        return {
            typeName: MintEvent.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(MintEvent.$typeName, ...[]),
            typeArgs: [],
            isPhantom: MintEvent.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => MintEvent.fromFields(fields),
            fromFieldsWithTypes: (item) => MintEvent.fromFieldsWithTypes(item),
            fromBcs: (data) => MintEvent.fromBcs(data),
            bcs: MintEvent.bcs,
            fromJSONField: (field) => MintEvent.fromJSONField(field),
            fromJSON: (json) => MintEvent.fromJSON(json),
            fromSuiParsedData: (content) => MintEvent.fromSuiParsedData(content),
            fromSuiObjectData: (content) => MintEvent.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return MintEvent.fetch(client, id); }),
            new: (fields) => {
                return new MintEvent([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return MintEvent.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(MintEvent.reified());
    }
    static get p() {
        return MintEvent.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("MintEvent", {
            typename: structs_1.TypeName.bcs,
            sui_amount_in: bcs_1.bcs.u64(),
            lst_amount_out: bcs_1.bcs.u64(),
            fee_amount: bcs_1.bcs.u64(),
        });
    }
    static fromFields(fields) {
        return MintEvent.reified().new({
            typename: (0, reified_1.decodeFromFields)(structs_1.TypeName.reified(), fields.typename),
            suiAmountIn: (0, reified_1.decodeFromFields)("u64", fields.sui_amount_in),
            lstAmountOut: (0, reified_1.decodeFromFields)("u64", fields.lst_amount_out),
            feeAmount: (0, reified_1.decodeFromFields)("u64", fields.fee_amount),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isMintEvent(item.type)) {
            throw new Error("not a MintEvent type");
        }
        return MintEvent.reified().new({
            typename: (0, reified_1.decodeFromFieldsWithTypes)(structs_1.TypeName.reified(), item.fields.typename),
            suiAmountIn: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.sui_amount_in),
            lstAmountOut: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.lst_amount_out),
            feeAmount: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.fee_amount),
        });
    }
    static fromBcs(data) {
        return MintEvent.fromFields(MintEvent.bcs.parse(data));
    }
    toJSONField() {
        return {
            typename: this.typename.toJSONField(),
            suiAmountIn: this.suiAmountIn.toString(),
            lstAmountOut: this.lstAmountOut.toString(),
            feeAmount: this.feeAmount.toString(),
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return MintEvent.reified().new({
            typename: (0, reified_1.decodeFromJSONField)(structs_1.TypeName.reified(), field.typename),
            suiAmountIn: (0, reified_1.decodeFromJSONField)("u64", field.suiAmountIn),
            lstAmountOut: (0, reified_1.decodeFromJSONField)("u64", field.lstAmountOut),
            feeAmount: (0, reified_1.decodeFromJSONField)("u64", field.feeAmount),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== MintEvent.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return MintEvent.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isMintEvent(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a MintEvent object`);
        }
        return MintEvent.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" || !isMintEvent(data.bcs.type)) {
                throw new Error(`object at is not a MintEvent object`);
            }
            return MintEvent.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return MintEvent.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching MintEvent object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isMintEvent(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a MintEvent object`);
            }
            return MintEvent.fromSuiObjectData(res.data);
        });
    }
}
exports.MintEvent = MintEvent;
MintEvent.$typeName = `${index_1.PKG_V1}::liquid_staking::MintEvent`;
MintEvent.$numTypeParams = 0;
MintEvent.$isPhantom = [];
/* ============================== RedeemEvent =============================== */
function isRedeemEvent(type) {
    type = (0, util_1.compressSuiType)(type);
    return type === `${index_1.PKG_V1}::liquid_staking::RedeemEvent`;
}
class RedeemEvent {
    constructor(typeArgs, fields) {
        this.__StructClass = true;
        this.$typeName = RedeemEvent.$typeName;
        this.$isPhantom = RedeemEvent.$isPhantom;
        this.$fullTypeName = (0, util_1.composeSuiType)(RedeemEvent.$typeName, ...typeArgs);
        this.$typeArgs = typeArgs;
        this.typename = fields.typename;
        this.lstAmountIn = fields.lstAmountIn;
        this.suiAmountOut = fields.suiAmountOut;
        this.feeAmount = fields.feeAmount;
    }
    static reified() {
        return {
            typeName: RedeemEvent.$typeName,
            fullTypeName: (0, util_1.composeSuiType)(RedeemEvent.$typeName, ...[]),
            typeArgs: [],
            isPhantom: RedeemEvent.$isPhantom,
            reifiedTypeArgs: [],
            fromFields: (fields) => RedeemEvent.fromFields(fields),
            fromFieldsWithTypes: (item) => RedeemEvent.fromFieldsWithTypes(item),
            fromBcs: (data) => RedeemEvent.fromBcs(data),
            bcs: RedeemEvent.bcs,
            fromJSONField: (field) => RedeemEvent.fromJSONField(field),
            fromJSON: (json) => RedeemEvent.fromJSON(json),
            fromSuiParsedData: (content) => RedeemEvent.fromSuiParsedData(content),
            fromSuiObjectData: (content) => RedeemEvent.fromSuiObjectData(content),
            fetch: (client, id) => __awaiter(this, void 0, void 0, function* () { return RedeemEvent.fetch(client, id); }),
            new: (fields) => {
                return new RedeemEvent([], fields);
            },
            kind: "StructClassReified",
        };
    }
    static get r() {
        return RedeemEvent.reified();
    }
    static phantom() {
        return (0, reified_1.phantom)(RedeemEvent.reified());
    }
    static get p() {
        return RedeemEvent.phantom();
    }
    static get bcs() {
        return bcs_1.bcs.struct("RedeemEvent", {
            typename: structs_1.TypeName.bcs,
            lst_amount_in: bcs_1.bcs.u64(),
            sui_amount_out: bcs_1.bcs.u64(),
            fee_amount: bcs_1.bcs.u64(),
        });
    }
    static fromFields(fields) {
        return RedeemEvent.reified().new({
            typename: (0, reified_1.decodeFromFields)(structs_1.TypeName.reified(), fields.typename),
            lstAmountIn: (0, reified_1.decodeFromFields)("u64", fields.lst_amount_in),
            suiAmountOut: (0, reified_1.decodeFromFields)("u64", fields.sui_amount_out),
            feeAmount: (0, reified_1.decodeFromFields)("u64", fields.fee_amount),
        });
    }
    static fromFieldsWithTypes(item) {
        if (!isRedeemEvent(item.type)) {
            throw new Error("not a RedeemEvent type");
        }
        return RedeemEvent.reified().new({
            typename: (0, reified_1.decodeFromFieldsWithTypes)(structs_1.TypeName.reified(), item.fields.typename),
            lstAmountIn: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.lst_amount_in),
            suiAmountOut: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.sui_amount_out),
            feeAmount: (0, reified_1.decodeFromFieldsWithTypes)("u64", item.fields.fee_amount),
        });
    }
    static fromBcs(data) {
        return RedeemEvent.fromFields(RedeemEvent.bcs.parse(data));
    }
    toJSONField() {
        return {
            typename: this.typename.toJSONField(),
            lstAmountIn: this.lstAmountIn.toString(),
            suiAmountOut: this.suiAmountOut.toString(),
            feeAmount: this.feeAmount.toString(),
        };
    }
    toJSON() {
        return Object.assign({ $typeName: this.$typeName, $typeArgs: this.$typeArgs }, this.toJSONField());
    }
    static fromJSONField(field) {
        return RedeemEvent.reified().new({
            typename: (0, reified_1.decodeFromJSONField)(structs_1.TypeName.reified(), field.typename),
            lstAmountIn: (0, reified_1.decodeFromJSONField)("u64", field.lstAmountIn),
            suiAmountOut: (0, reified_1.decodeFromJSONField)("u64", field.suiAmountOut),
            feeAmount: (0, reified_1.decodeFromJSONField)("u64", field.feeAmount),
        });
    }
    static fromJSON(json) {
        if (json.$typeName !== RedeemEvent.$typeName) {
            throw new Error("not a WithTwoGenerics json object");
        }
        return RedeemEvent.fromJSONField(json);
    }
    static fromSuiParsedData(content) {
        if (content.dataType !== "moveObject") {
            throw new Error("not an object");
        }
        if (!isRedeemEvent(content.type)) {
            throw new Error(`object at ${content.fields.id} is not a RedeemEvent object`);
        }
        return RedeemEvent.fromFieldsWithTypes(content);
    }
    static fromSuiObjectData(data) {
        if (data.bcs) {
            if (data.bcs.dataType !== "moveObject" || !isRedeemEvent(data.bcs.type)) {
                throw new Error(`object at is not a RedeemEvent object`);
            }
            return RedeemEvent.fromBcs((0, utils_1.fromB64)(data.bcs.bcsBytes));
        }
        if (data.content) {
            return RedeemEvent.fromSuiParsedData(data.content);
        }
        throw new Error("Both `bcs` and `content` fields are missing from the data. Include `showBcs` or `showContent` in the request.");
    }
    static fetch(client, id) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const res = yield client.getObject({ id, options: { showBcs: true } });
            if (res.error) {
                throw new Error(`error fetching RedeemEvent object at id ${id}: ${res.error.code}`);
            }
            if (((_b = (_a = res.data) === null || _a === void 0 ? void 0 : _a.bcs) === null || _b === void 0 ? void 0 : _b.dataType) !== "moveObject" ||
                !isRedeemEvent(res.data.bcs.type)) {
                throw new Error(`object at id ${id} is not a RedeemEvent object`);
            }
            return RedeemEvent.fromSuiObjectData(res.data);
        });
    }
}
exports.RedeemEvent = RedeemEvent;
RedeemEvent.$typeName = `${index_1.PKG_V1}::liquid_staking::RedeemEvent`;
RedeemEvent.$numTypeParams = 0;
RedeemEvent.$isPhantom = [];
