// @suilend/sdk/parsers/lendingMarket.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseLendingMarket = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const rateLimiter_1 = require("./rateLimiter");
const reserve_1 = require("./reserve");
const parseLendingMarket = (lendingMarket, reserves, coinMetadataMap, nowS, lendingMarketMetadata) => {
    var _a, _b, _c;
    const id = lendingMarket.id;
    const type = lendingMarket.$typeArgs[0];
    const version = lendingMarket.version;
    const parsedReserves = reserves.map((reserve) => (0, reserve_1.parseReserve)(reserve, coinMetadataMap));
    const obligations = lendingMarket.obligations;
    const parsedRateLimiter = (0, rateLimiter_1.parseRateLimiter)(lendingMarket.rateLimiter, nowS);
    const feeReceiver = lendingMarket.feeReceiver;
    const badDebtUsd = new bignumber_js_1.default(lendingMarket.badDebtUsd.value.toString());
    const badDebtLimitUsd = new bignumber_js_1.default(lendingMarket.badDebtLimitUsd.value.toString());
    // Custom
    let depositedAmountUsd = new bignumber_js_1.default(0);
    let borrowedAmountUsd = new bignumber_js_1.default(0);
    let tvlUsd = new bignumber_js_1.default(0);
    parsedReserves.forEach((parsedReserve) => {
        depositedAmountUsd = depositedAmountUsd.plus(parsedReserve.depositedAmountUsd);
        borrowedAmountUsd = borrowedAmountUsd.plus(parsedReserve.borrowedAmountUsd);
        tvlUsd = tvlUsd.plus(parsedReserve.availableAmountUsd);
    });
    return {
        version,
        reserves: parsedReserves,
        obligations,
        rateLimiter: parsedRateLimiter,
        feeReceiver,
        badDebtUsd,
        badDebtLimitUsd,
        depositedAmountUsd,
        borrowedAmountUsd,
        tvlUsd,
        // Metadata
        id,
        type,
        ownerCapId: (_a = lendingMarketMetadata === null || lendingMarketMetadata === void 0 ? void 0 : lendingMarketMetadata.lendingMarketOwnerCapId) !== null && _a !== void 0 ? _a : undefined,
        name: (_b = lendingMarketMetadata === null || lendingMarketMetadata === void 0 ? void 0 : lendingMarketMetadata.name) !== null && _b !== void 0 ? _b : type.split("::").slice(1).join(" "),
        isHidden: (_c = lendingMarketMetadata === null || lendingMarketMetadata === void 0 ? void 0 : lendingMarketMetadata.isHidden) !== null && _c !== void 0 ? _c : true,
        /**
         * @deprecated since version 1.0.3. Use `depositedAmountUsd` instead.
         */
        totalSupplyUsd: depositedAmountUsd,
        /**
         * @deprecated since version 1.0.3. Use `borrowedAmountUsd` instead.
         */
        totalBorrowUsd: borrowedAmountUsd,
    };
};
exports.parseLendingMarket = parseLendingMarket;
